programs/hyperplane/src/curve/stable.rs

```diff
@@ -0,0 +1,631 @@
+//! The stableswap invariant calculator.
+use crate::curve::calculator::DynAccountSerialize;
+use crate::require_msg;
+use crate::state::StableCurve;
+use anchor_lang::Result;
+use {
+    crate::{
+        curve::calculator::{
+            CurveCalculator, RoundDirection, SwapWithoutFeesResult, TradeDirection,
+            TradingTokenResult,
+        },
+        error::SwapError,
+    },
+    spl_math::{checked_ceil_div::CheckedCeilDiv, precise_number::PreciseNumber, uint::U256},
+    std::convert::TryFrom,
+};
+
+const N_COINS: u8 = 2;
+const N_COINS_SQUARED: u8 = 4;
+const ITERATIONS: u8 = 32;
+
+/// Minimum amplification coefficient.
+pub const MIN_AMP: u64 = 1;
+
+/// Maximum amplification coefficient.
+pub const MAX_AMP: u64 = 1_000_000;
+
+/// Calculates A for deriving D
+///
+/// Per discussion with the designer and writer of stable curves, this A is not
+/// the same as the A from the whitepaper, it's actually `A * n**(n-1)`, so when
+/// you set A, you actually set `A * n**(n-1)`. This is because `D**n / prod(x)`
+/// loses precision with a huge A value.
+///
+/// There is little information to document this choice, but the original contracts
+/// use this same convention
+fn compute_a(amp: u64) -> Option<u64> {
+    amp.checked_mul(N_COINS as u64)
+}
+
+/// Returns self to the power of b
+fn checked_u8_power(a: &U256, b: u8) -> Option<U256> {
+    let mut result = *a;
+    for _ in 1..b {
+        result = result.checked_mul(*a)?;
+    }
+    Some(result)
+}
+
+/// Returns self multiplied by b
+fn checked_u8_mul(a: &U256, b: u8) -> Option<U256> {
+    let mut result = *a;
+    for _ in 1..b {
+        result = result.checked_add(*a)?;
+    }
+    Some(result)
+}
```

Comment:
What's wrong with `checked_mul`? A for loop looks very badly optimized...
```suggestion
fn checked_u8_mul(a: U256, b: u8) -> Option<U256> {
    let b = U256::from(b);
    a.checked_mul(b)
}
```
Bonus: why take a reference as a parameter since you dereference it immediately?
Also, it's probably not as easy to demonstrate but this is also true for `pow` IMO.

Reply:
Thanks for the suggestion - I agree they seem badly optimised but I have no data on compute unit usage until I add BPF tests. 

I assume the original authors worked this out to be more efficient under BPF somehow - but I will come back to this once the tests are added.